### ISSUE TITLE
UI rights do not check for user resolver filters

### ISSUE THREAD

**Issue by quynh-axiadids** [privacyidea/privacyidea#1496]
# What did you try to do?

Display UI rights correctly according to policies and their filters.
For example:
In the same realm realmA, there are 2 resolvers resolverX and resolverY. In policies setup, resolverX has assign action enabled, whereas resolverY does not have assign action enabled.

# What outcome did you expect?
When logging to the UI, users in resolverX should see assign option in the menu. Users in resolverY should not see assign option in the UI.

# What outcome did you experience?

Users in both resolvers of realmA see the assign option. Although when actually self-assigning a token, it will fail for user in resolverY.

# Configuration

* privacyIDEA Version: 2.23.3

* Installation method: from source

## more details: 

* OS: Amazon Linux

* Webserver: Nginx

* Tokendatabase: PostgreSQL

# Log file

N/A

# Side notes:

I checked this method and it seems that we do not pass the resolver info when checking for UI rights, and may result in the described outcome?

```
# privacyidea/privacyidea/lib/policy.py
def ui_get_rights(self, scope, realm, username, client=None):
[...]
```
This is not a very critical bug. However, it affects user experience and causes some confusion. It would be nice to have it adjusted!

Cheers,

Quynh Nguyen

**Comment by cornelinux** [comment]
Well, works as programmed! ;-)
Honestly there have been several ocasions, where we skipped the resolvers. Especially the login can be a bit tricky, since the user is not available before he authenticated.

But in this case you are right!
It happens here: https://github.com/privacyidea/privacyidea/blob/master/privacyidea/lib/policy.py#L761

We should change it like this:

~~~~python
        if scope == SCOPE.ADMIN:
            adminrealm = realm
            logged_in_user["role"] = ROLE.ADMIN
            if adminrealm: # internal admins can not be resolved
                user = User(username, adminrealm)
                resolver = user.resolver
        elif scope == SCOPE.USER:
            userrealm = realm
            logged_in_user["role"] = ROLE.USER
            resolver = User(username, userrealm).resolver
~~~~
Then we can use the ``resolver`` in the get_policies.

**Comment by darrenche** [comment]
Hello,

I noticed this issue was fixed and is present inside the tag **v3.3dev1-diagrams**. We were wondering why this change has since been removed and is not present inside master. 

Thank you for your time!

**Comment by cornelinux** [comment]
It has not been removed, it has been changed. Or the policy handling has been changed.
You can see, that the test for different resolvers still works out:
https://github.com/privacyidea/privacyidea/blob/master/tests/test_lib_policy.py#L623

It has changed this way, that not (user, realm) is passed to find the policies but actualy the user_object (which contiains the resolver) is passed.
https://github.com/privacyidea/privacyidea/blob/master/privacyidea/lib/policy.py#L974

The whole policy handling is moving to a top level matching mechanism (``Match``), which allows us to add new policy conditions centrally more easily.

Do you run a very uncurrent fork? This is unfortunate, since it is probably very difficult -> impossible to merge this.

If you are from California you may want to listen to a talk of mine at scale18x.
https://www.socallinuxexpo.org/scale/18x/presentations/flexible-reliable-and-lasting-multi-factor-authentication-privacyidea (Interesting place to drop some ad)
